### ISSUE TITLE
Should we nest variables in s-strings?

### ISSUE THREAD

**Issue by max-sixty** [PRQL/prql#1155]
> To me it seems that while functions are defined through s-string substitutions, they are essentially opaque to the compiler so there's not much it can do to ensure semantic correctness.

Yes I think this is spot-on. We're using literal replacement and expecting semantic correctness. If we use semantic replacement, it works — i.e. `func fun3 n d -> n / d`.

---

I'm not sure of the best solution — the current design is correct is the literal sense, but at least with these examples, it's clearly not doing what the user intends.

Potential reasons to not wrap / nest the variable:
- Are there cases where nesting the variable would be wrong? Seems unlikely at first glance.
- I'm hesitate to have an additional parsing stage to remove superfluous parentheses

Alternatively, is there a simple heuristic we could use to decide whether to nest? e.g. "has non-word characters" (though this makes the special-cases worse, not better)

_Originally posted by @max-sixty in https://github.com/prql/prql/issues/1116#issuecomment-1323588180_

**Comment by richb-hanover** [comment]
I also agree with @snth in #1116: the PRQL compiler cannot see inside the s-string. 

_Updated this paragraph..._ I'm not so worried about leaving "extra paren's" around each function parameter's expansion. It would be correct (if we believe that after further testing), easily explicable, and mostly not "in the way". (The use case of people reading the generated SQL to learn PRQL will still work. Once you start using functions heavily, you've internalized most of how PRQL works.) 

So... I think we agree that adding paren's to my numeric parameter example would "do the right thing". But what about string parameters? For example, the URL in the book:

```
select url = f"http{tls}://www.{domain}.{tld}/{page}"
```
turns into:

```
SELECT
  CONCAT(
    'http',
    tls,
    '://www.',
    domain,
    '.',
    tld,
    '/',
    page
  ) AS url
FROM
  web
```
In a one-out-of-one test, I used [Postgres Online](https://extendsclass.com/postgresql-online.html) to test that example (I substituted "tls", "domain", etc names for columns in the "scientist" table), then ran the query. It worked as expected. Sample query below:

I then wrapped those parameters in `(...)` and the query also succeeded. One out of one! Ship it! :-)

```
select CONCAT(
    'http',
    (id),
    '://www.',
    (firstname),
    '.',
    (lastname),
    'Hi Rich'
  ) AS url
FROM
  scientist;
```

**Comment by aljazerzen** [comment]
> the PRQL compiler cannot see inside the s-string.

That's the essence of it. 

Compiler treats s-strings as a blackbox expression with unknown type. They are an escape hatch for accessing SQL underneath and when you use them, you take the responsibility for producing correct SQL.

@max-sixty I do have a counter example for adding parenthesis:

```elm
func set_schema_and_select column -> s"; SET search_path TO 'my_schema'; SELECT {column}"

from employees
select [set_schema_and_select first_name, last_name]
```
... which compiles to:
```sql
SELECT
;
SET
  search_path TO 'my_schema';
SELECT
  first_name,
  last_name
FROM
  employees
```

I think that we should treat s-strings as `unsafe` in Rust. Heavily discouraged, but available as a last resort.

**Comment by max-sixty** [comment]
Great example @aljazerzen 


> I think that we should treat s-strings as `unsafe` in Rust. Heavily discouraged, but available as a last resort.

✔️

**Comment by max-sixty** [comment]
> Great example @aljazerzen
> 
> > I think that we should treat s-strings as `unsafe` in Rust. Heavily discouraged, but available as a last resort.
> 
> ✔️

So I would vote to not proceed with this at the moment; instead we should invest in minimizing the need for s-strings. They'll always be available, but will at times require active attention from the user.

**Comment by richb-hanover** [comment]
Let me see if I understand the current state of the discussion, with an eye toward documenting it to avoid others from stumbling as I have.

The proposal is:

* Leave S-strings and F-strings as-is, and put the responsibility for correctness to the user. 
* As we learn of new awkward cases, PRQL can add more facilities/functions to the language that it _does_ check.

I agree: this is reasonable. Here's proposed wording to document the decision. Feel free to edit (or reject):

> Warning: The PRQL compiler simply places a literal copy of each variable into the resulting string without further checking. Considerations:
>
> * Because F-strings and S-strings are opaque, the PRQL compiler cannot do any checking. You should ensure that the resulting SQL is what you expect.
> * Let us know if you encounter cases where it is difficult to generate the SQL that you need. We will consider extending PRQL to handle the case.

---
PS How did we find ourselves in this rabbit hole? Here's what I remember:

I wanted to make a "percent-increase" function. `100*( new - old ) / old`. But that didn't give the correct result if it was all integer data. So I created a S-string to CAST() the numerator. But this lead to an incorrect result if the denominator was, itself, a quotient... So paren's around the denominator worked, but that felt alarming, so I wrote up the report.

BUT... PRQL now passes full decimal values through, so I can write `100.0 * (new - old)/old` No need for S-string or CAST() - the PRQL compiler handles it all.

Thanks for listening
